action.yml

```diff
@@ -42,7 +47,25 @@ runs:
           echo "unknown platform $platform"
           exit 1
         }
-        Invoke-WebRequest https://cygwin.com/setup-$platform.exe -OutFile C:\setup.exe
+        $setupFileName = "setup-$platform.exe"
+        Invoke-WebRequest "https://cygwin.com/$setupFileName" -OutFile C:\setup.exe
+
+        if ('${{ inputs.check-hash }}'.ToLower() -in @('','true', 'yes', '1', 'on')) {
+          $actualHash = $(Get-FileHash -LiteralPath C:\setup.exe -Algorithm SHA512).Hash
+          $expectedHashLines = $(Invoke-WebRequest -Uri https://cygwin.com/sha512.sum).ToString() -split "`n"
+          foreach ($expectedHashLine in $expectedHashLines) {
+            if ($expectedHashLine -match "^(\S+)\s+(\S+)$") {
+              $expectedHash = $matches[1]
+              $expectedFileName = $matches[2]
+              if ($expectedFileName -ieq $setupFileName) {
+                if ($expectedHash -ine $actualHash) {
+                  throw "Invalid hash of the downloaded setup!`nExpected: $expectedHashLine`nActual  : $actualHash  $expectedFileName"
+                }
+                Write-Host "The downloaded file has the expected hash ($actualHash)"
+              }
```

Comment:
It seems like this silently treats "filename isn't in the checksum file" as a success, which perhaps not what's wanted.

Reply:
I've updated the code: it now prints a warning if the name of the setup file is not present in the downloaded sha512.sum file